compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala

```diff
@@ -180,14 +156,78 @@ class PickleQuotes extends MacroTransform {
 
     val holeMaker = new HoleContentExtractor
     val body1 = holeMaker.transform(quote.body)
-    val body2 =
-      if quote.isTypeQuote then body1
-      else Inlined(Inlines.inlineCallTrace(ctx.owner, quote.sourcePos), Nil, body1)
-    val quote1 = cpy.Quote(quote)(body2)
+    val quote1 = cpy.Quote(quote)(body1, quote.tags)
 
     (holeMaker.getContents(), quote1)
   end makeHoles
 
+  /** Encode quote tags as holes in the quote body.
+   *
+   *  ```scala
+   *    '<t, u>{ ... t.Underlying ... u.Underlying ... }
+   *  ```
+   *  becomes
+   * ```scala
+   *  '<t, u>{
+   *    type T = {{ 0 | .. | .. | .. }}
+   *    type U = {{ 1 | .. | .. | .. }}
+   *    ... T ... U ...
+   *  }
+   * ```
+   */
+  private def encodeTypeArgs(quote: tpd.Quote)(using Context): tpd.Quote =
+    if quote.tags.isEmpty then quote
+    else
+      val tdefs = quote.tags.zipWithIndex.map(mkTagSymbolAndAssignType)
+      val typeMapping = quote.tags.map(_.tpe).zip(tdefs.map(_.symbol.typeRef)).toMap
+      val typeMap = new TypeMap {
+        override def apply(tp: Type): Type = tp match
+          case TypeRef(tag: TermRef, _) if tp.typeSymbol == defn.QuotedType_splice =>
+            typeMapping.getOrElse(tag, tp)
+          case _ => mapOver(tp)
+      }
+      def treeMap(tree: Tree): Tree = tree match
+          case Select(qual, _) if tree.symbol == defn.QuotedType_splice =>
+            typeMapping.get(qual.tpe) match
+              case Some(tag) => TypeTree(tag).withSpan(tree.span)
+              case None => tree
+          case _ => tree
+      val body1 = new TreeTypeMap(typeMap, treeMap).transform(quote.body)
+      cpy.Quote(quote)(Block(tdefs, body1), quote.tags)
+
+  private def mkTagSymbolAndAssignType(typeArg: Tree, idx: Int)(using Context): TypeDef = {
+    val holeType = getTypeHoleType(typeArg.tpe.select(tpnme.Underlying))
+    val hole = cpy.Hole(typeArg)(isTerm = false, idx, Nil, EmptyTree, TypeTree(holeType))
+    val local = newSymbol(
+      owner = ctx.owner,
+      name = UniqueName.fresh(hole.tpe.dealias.typeSymbol.name.toTypeName),
+      flags = Synthetic,
+      info = TypeAlias(typeArg.tpe.select(tpnme.Underlying)),
+      coord = typeArg.span
+    ).asType
+    local.addAnnotation(Annotation(defn.QuotedRuntime_SplicedTypeAnnot, typeArg.span))
+    ctx.typeAssigner.assignType(untpd.TypeDef(local.name, hole), local).withSpan(typeArg.span)
+  }
+
+  /** Remove references to local types that will not be defined in this quote */
+  private def getTypeHoleType(using Context) = new TypeMap() {
```

Comment:
Looks like this code already existed before but FWIW, I think TypeOps.AvoidMap would be better suited here (or its superclass ApproximatingTypeMap if more flexibility is needed), just using TypeMap doesn't take variance and bounds into account.